### PR TITLE
Support deriving Functor, Foldable, and Traversable

### LINKED ISSUE #184: Support monads
We're surprisingly close to being able to support monads (and, with them, list comprehensions). Singletonizing the basic class definitions works. But there's danger on the horizon.

The problem has all to do with CUSKs. As `Note [CUSKification]` in Promote describes, we have little control over when GHC infers a CUSK. And this causes havoc when we want GHC to infer the kind of higher-kinded type variables. Specifically, the `PFunctor` class has a CUSK, and so the user needs to write `class Functor (f :: * -> *)`. Putting in the kind annotations here isn't so, so terrible.

What's worse is that we run into similar problems elsewhere. I haven't tracked down the precise details, but compiling Base.hs on the `monads` branch led to an error where a defunctionalization symbol has a CUSK, and thus inferring that a kind variable should be higher-kinded fails.

Short of fixing CUSKs in GHC, the solution might be to emit GADT-style declarations of defunctionalization symbols instead of H98-style. This would allow us finer control of whether we have a CUSK. (Recall that, with `TypeInType`, all kind variables must be explicitly quantified in order to have a CUSK. So we just won't do that.) This is tantalizingly close. But I'm not going to get there tonight.

Checklist:

* [x] Add `Functor`, `Applicative`, `Monad`, `Alternative`, `MonadPlus`, and `MonadZip` classes, as well as supporting functions. (https://github.com/goldfirere/singletons/pull/314)
* [x] Add `Identity`, `Const`, `Foldable`, and `Traversable`, as well as supporting functions. (https://github.com/goldfirere/singletons/pull/337)
* [ ] Add support for deriving `Functor`, `Foldable`, and `Traversable`. (#344)

### PR DESCRIPTION
**Pull request by RyanGlScott**
This implements the last bullet point in #184: supporting `DeriveFunctor`, `DeriveFoldable`, and `DeriveTraversable` in `singletons`. Much of the code in this commit is cargo-culted from the GHC source code, since the algorithms that it employs to derive those classes work much the same in `singletons`.

Fixes #184 for good.

### CODE REVIEW

File: src/Data/Singletons/Partition.hs [hunk 1]
```diff
@@ -224,6 +235,18 @@ partitionDeriving mb_strat deriv_pred mb_ctxt ty data_decl =
        , deriv_name == enumName
       -> mk_derived_inst <$> mk_instance mkEnumInstance
 
+       | stock_or_default
```
Comment (goldfirere) [inline src/Data/Singletons/Partition.hs]: This region could be cleaned up a bit by having an mapping (or an associative list, at list) from class names to instance makers.

File: src/Data/Singletons/Prelude/Functor.hs [hunk 2]
```diff
@@ -194,13 +195,8 @@ $(singletonsOnly [d|
...
   void x = () <$ x
 
-  -- deriving instance Functor ((,) a)
-  instance Functor ((,) a) where
-      fmap f (x,y) = (x, f y)
-
-  -- deriving instance Functor Down
-  instance Functor Down where
-    fmap f (Down x) = Down (f x)
+  deriving instance Functor ((,) a)
```
Comment (goldfirere) [inline src/Data/Singletons/Prelude/Functor.hs]: Why are these here? Maybe there are module dependencies around `Down`, but the tuple instance should be in `Monad.Internal`, no?

File: src/Data/Singletons/Prelude/Identity.hs [hunk 3]
```diff
@@ -81,9 +83,7 @@ $(singletonsOnly [d|
     showsPrec d (Identity x) = showParen (d > 10) $
       showString "Identity " . showsPrec 11 x
 
-  -- deriving instance Functor Identity
-  instance Functor Identity where
-    fmap f (Identity x) = Identity (f x)
+  deriving instance Functor Identity
```
Comment (goldfirere) [inline src/Data/Singletons/Prelude/Identity.hs]: `Foldable`? `Traversable`?

File: src/Data/Singletons/Prelude/Monoid.hs [hunk 4]
```diff
@@ -178,9 +179,7 @@ $(singletonsOnly [d|
     pure = First . pure
     First f <*> First x = First (f <*> x)
 
-  -- deriving instance Functor First
-  instance Functor First where
-    fmap f (First x) = First (fmap f x)
+  deriving instance Functor First
```
Comment (goldfirere) [inline src/Data/Singletons/Prelude/Monoid.hs]: `Traversable`?

File: src/Data/Singletons/Prelude/Monoid.hs [hunk 5]
```diff
@@ -198,9 +197,7 @@ $(singletonsOnly [d|
     pure = Last . pure
     Last f <*> Last x = Last (f <*> x)
 
-  -- deriving instance Functor Last
-  instance Functor Last where
-    fmap f (Last x) = Last (fmap f x)
+  deriving instance Functor Last
```
Comment (goldfirere) [inline src/Data/Singletons/Prelude/Monoid.hs]: `Traversable`?

File: src/Data/Singletons/Prelude/Semigroup.hs [hunk 6]
```diff
@@ -227,9 +206,7 @@ $(singletonsOnly [d|
     enumFromTo (First a) (First b) = First `map` enumFromTo a b
     enumFromThenTo (First a) (First b) (First c) = First `map` enumFromThenTo a b c
 
-  -- deriving instance Functor First
-  instance Functor First where
-    fmap f (First x) = First (f x)
+  deriving instance Functor First
```
Comment (goldfirere) [inline src/Data/Singletons/Prelude/Semigroup.hs]: This isn't the first time I've seen this declaration...

File: src/Data/Singletons/Prelude/Semigroup/Internal.hs [hunk 7]
```diff
@@ -178,9 +177,7 @@ $(singletonsOnly [d|
     pure = Sum
     Sum f <*> Sum x = Sum (f x)
 
-  -- deriving instance Functor Sum
-  instance Functor Sum where
-    fmap f (Sum x) = Sum (f x)
+  deriving instance Functor Sum
```
Comment (goldfirere) [inline src/Data/Singletons/Prelude/Semigroup/Internal.hs]: `Foldable`? `Traversable`?

File: src/Data/Singletons/Partition.hs (hunk 1 above)
Comment (RyanGlScott) [reply src/Data/Singletons/Partition.hs]: Good idea, I'll do that.

File: src/Data/Singletons/Prelude/Functor.hs (hunk 2 above)
Comment (RyanGlScott) [reply src/Data/Singletons/Prelude/Functor.hs]: I suppose so. I'll try moving the `(,)` instances to `Monad.Internal` and see if anything blows up.

File: src/Data/Singletons/Prelude/Identity.hs (hunk 3 above)
Comment (RyanGlScott) [reply src/Data/Singletons/Prelude/Identity.hs]: There's a good reason for this one—`Data.Foldable` and `Data.Traversable` must import `Identity`, `Const`, `Sum`, `Product`, `First`, `Last`, and several other newtypes to implement some of its functionality. (For similar reasons, `base` also employs a similar module structure.)

File: src/Data/Singletons/Prelude/Monoid.hs (hunk 4 above)
Comment (RyanGlScott) [reply src/Data/Singletons/Prelude/Monoid.hs]: See my comment above.

File: src/Data/Singletons/Prelude/Monoid.hs (hunk 5 above)
Comment (RyanGlScott) [reply src/Data/Singletons/Prelude/Monoid.hs]: See my comment above.

File: src/Data/Singletons/Prelude/Semigroup.hs (hunk 6 above)
Comment (RyanGlScott) [reply src/Data/Singletons/Prelude/Semigroup.hs]: Note that the [`First` in `Data.Semigroup`](https://hackage.haskell.org/package/base-4.11.1.0/docs/Data-Semigroup.html#t:First) is not the same [`First` as in `Data.Monoid`](https://hackage.haskell.org/package/base-4.11.1.0/docs/Data-Monoid.html#t:First).

File: src/Data/Singletons/Prelude/Semigroup/Internal.hs (hunk 7 above)
Comment (RyanGlScott) [reply src/Data/Singletons/Prelude/Semigroup/Internal.hs]: See my comment above.

**Comment by goldfirere** [comment]
OK. I didn't do a close analysis. If you're happy with the instances that are declared, then so am I.

File: src/Data/Singletons/Prelude/Functor.hs (hunk 2 above)
Comment (RyanGlScott) [reply src/Data/Singletons/Prelude/Functor.hs]: Ugh, _now_ I remember why I didn't do this before. It's because these instances use `Monoid`, but `Data.Singletons.Prelude.Monad.Internal` occurs much lower in the dependency graph than `Data.Singletons.Prelude.Monoid`. It would be possible to do even more internal module reorganization to allow this, but I'm not sure it's worth it.

**Comment by goldfirere** [comment]
OK. Don't let me make busywork for you. Do what you think is best.
